### DISCUSSION TITLE
Error for output_centroids

### DISCUSSION THREAD

**Discussion by Clark-domain** [idaholab/moose#17965] [Q&A General] [answered]
Hi,  everyone,

I am trying to get the centroids of the grains by FeatureVolumeVectorPostprocessor. I followed the MOOSE instruction and added execute_on = 'timestep_begin' & output_centroids = true. But it did not work and showed the error as below.  How to solve it?

  [./grain_volumes] 
    type = FeatureVolumeVectorPostprocessor
    flood_counter = grain_center
    execute_on = 'timestep_begin'
    output_centroids = true 
  [../]

![1](https://user-images.githubusercontent.com/74393696/120091609-b8c12000-c0da-11eb-880e-10af042e3056.jpg)

**Comment by aeslaughter** [comment]
This could be a bug. The error indicates that the output you desire already exists. What input block is your snippet in: [Postprocessors] or [UsersObjects]?

**Reply by Clark-domain** [reply]
The input file that I modified was from (modules/phase_field/test/tests/rigidbodymotion/grain_motion.i).
I tried either [Postprocessors] or [UsersObjects] but neither worked.
What could be possible to get the output for Centroids? Thanks very much.

[Postprocessors] #or [UserObjects]
  [./grain_center]
    type = GrainTracker
    variable = eta
    outputs = none
    compute_var_to_feature_map = true
    execute_on = 'timestep_begin'
  [../]
[]

[VectorPostprocessors]
  [./grain_volumes]
    type = FeatureVolumeVectorPostprocessor
    flood_counter = grain_center
    execute_on = 'timestep_begin'
    output_centroids = true
  [../]
[]

**Reply by aeslaughter** [reply]
Can you provide a minimum input file that show the error?

**Reply by Clark-domain** [reply]
To keep it short, here is the input file for test purposes. You will need to enable PhaseField Module to run it. 
Many thanks. 

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 25
  ny = 15
  nz = 0
  xmax = 50
  ymax = 25
  zmax = 0
  elem_type = QUAD4
[]

[Variables]
  [./c]
    order = FIRST
    family = LAGRANGE
  [../]
  [./w]
    order = FIRST
    family = LAGRANGE
  [../]
  [./eta]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./c_res]
    type = SplitCHParsed
    variable = c
    f_name = F
    kappa_name = kappa_c
    w = w
    args = eta
  [../]
  [./w_res]
    type = SplitCHWRes
    variable = w
    mob_name = M
  [../]
  [./time]
    type = CoupledTimeDerivative
    variable = w
    v = c
  [../]

  [./eta_dot]
    type = TimeDerivative
    variable = eta
  [../]
  [./acint_eta]
    type = ACInterface
    variable = eta
    mob_name = M
    args = c
    kappa_name = kappa_eta
  [../]
  [./acbulk_eta]
    type = AllenCahn
    variable = eta
    mob_name = M
    f_name = F
    args = c
  [../]
[]

[Materials]
  [./pfmobility]
    type = GenericConstantMaterial
    prop_names = 'M    kappa_c  kappa_eta'
    prop_values = '5.0  2.0      0.1'
  [../]
  [./free_energy]
    type = DerivativeParsedMaterial
    args = 'c eta'
    constant_names = 'barr_height  cv_eq'
    constant_expressions = '0.1          1.0e-2'
    function = 16*barr_height*(c-cv_eq)^2*(1-cv_eq-c)^2+(c-eta)^2
    derivative_order = 2
  [../]
[]

[Postprocessors]
  [./grain_center]
    type = GrainTracker
    variable = eta
    outputs = none
    compute_var_to_feature_map = true
    execute_on = 'timestep_begin'
  [../]

[]

[VectorPostprocessors]
  [./grain_volumes]
    type = FeatureVolumeVectorPostprocessor
    flood_counter = grain_center
    execute_on = 'timestep_begin'
    output_centroids = true
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  dt = 0.2
  num_steps = 4
[]

[Outputs]
  exodus = true
[]

[ICs]
  [./rect_c]
    y2 = 20.0
    y1 = 5.0
    inside = 1.0
    x2 = 30.0
    variable = c
    x1 = 10.0
    type = BoundingBoxIC
  [../]
  [./rect_eta]
    y2 = 20.0
    y1 = 5.0
    inside = 1.0
    x2 = 30.0
    variable = eta
    x1 = 10.0
    type = BoundingBoxIC
  [../]
[]
```

**Reply by GiudGiud** [reply]
Any news on this @aeslaughter @KaijieNing ?

**Reply by aeslaughter** [reply]
No. I need to try this out. I wonder if @permcody or @dschwen has an insight, since this if from the grain tracking system?

**Reply by tonkmr** [reply]
I am getting the same error and I have several projects that need to use this capability. Has anyone looked into it yet @aeslaughter @permcody @dschwen ?

**Reply by dschwen** [reply]
Yeah, clearly a bug. The `declareVector("centroid_x");` call is in `FeatureVolumeVectorPostprocessor::execute()` instead of in the constructor. That causes it to be redeclared over and over.

**Comment by dschwen** [comment, marked as answer]
Fix is up at https://github.com/idaholab/moose/pull/18672

**Reply by ykvishal** [reply]
Is this working for anyone? I am getting following error:

> *** ERROR ***
The following error occurred in the object "grain_volumes", of type "FeatureVolumeVectorPostprocessor".

>While declaring a VectorPostprocessor value with the name "centroid_x",
a Reporter with the same name has already been declared.

>VectorPostprocessor "grain_volumes/centroid_x":
  Type:
    std::vector<double>
  Producer:
    FeatureVolumeVectorPostprocessor "grain_volumes"
  Context type:
    ReporterGeneralContext<std::vector<double> >
  Consumer(s):
    None

I installed MOOSE last week and my input file is
[grain_growth_2D_graintrackerwithCentroid.i.txt](https://github.com/idaholab/moose/files/7498501/grain_growth_2D_graintrackerwithCentroid.i.txt)


Thank you.

Vishal

**Reply by Clark-domain** [reply]
Your input file was good. I was able to run it and had all the outputs. Best.

**Reply by ykvishal** [reply]
Thanks for the confirmation! Then something strange is happening my moose installation.

**Reply by Clark-domain** [reply]
You can update your MOOSE and try it again.

**Reply by ykvishal** [reply]
Yes, I installed MOOSE last week. However, I also added one old PR (related to grain growth initial condition) prior to make command, and it may have caused some problem. I will try again without this PR and see what happens. Thanks for the suggestion.
